### ISSUE TITLE
High Level Expressions

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#7933]
### Summary

We should make space for high level query optimization.  There are a couple of ways to do this.  This issue includes motivation, a description of two approaches, and some thoughts on trade-offs.

### Motivation

There are a variety of situations where we might want to rewrite a user's code.

#### Dataframes

1. Column projection, as in `dd.read_parquet(...)["x"]` -> `dd.read_parquet(..., columns=["x"])`
2. Predicate pushdown (same as above)
3. High level expression fusion (what we do today with blockwise)
4. Pushing length calls down through elementwise calls
5. Pushing filters earlier in a computation
6. ...

#### Arrays

1.  Automatic rechunking at the beginning of a computation based on the end of the computation
2. Slicing

### History

Today there is no real place where we capture a user's intent or their lines of code.  We immediately create a task graph for the requested operation, dump it into a mapping, and create a new dask.dataframe.DataFrame or dask.array.Array instance.  That instance has no knowledge of what created it, or what created the other input dataframes on which it depends.

This was a good choice early on.  It made it easy for us to quickly implement lots of complex operations without thinking about a class hierarchy for them.  This choice followed on from the choices of Blaze, where we started with high level expressions, but got a bit stuck because they constrained our thinking (and no one really cares about high level query optimization for a system that they don't use.

However today we have maybe reached a point where our "keep everything low-level and simple" strategy has hit a limit, and now we're curious about how best to bolt on a high level expression system.  Doing this smoothly given the current system is hard.  I see two ways out.

### High Level Graph layers

We *do* have a record of what operations came before us in the high level graph layers.  Currently the API of layers is very generic.  They must be a mapping that adheres to the Dask graph spec, and they must be serializable in a certain way.  There are some consistent subclasses, like blockwise, which enable high level optimizations which have proven useful.  

There isn't really much structure here though, and as a result it's hard to do interesting optimizations.  For example it would be nice if we could change a layer at the very bottom of the graph, and then replay all of the operations on that input over again to see how they would change.  High level layers today don't have enough shared structure that we know how to do this.

I like High Level Graph Layers because it gives us a space to hijack and add in all sorts of complex machinery, but without affecting the user-facing Dataframe class.  We would have to add a lot more structure here though, and we'll always be working around the collection class, which is a drawback.

### Collection subclasses

I'm going to focus on an alternative that is a bit more radical.  We could also have every user call generate a DataFrame subclass.  There would still be a DataFrame instance that took in a generic graph/divisions/meta, but that would be mostly for backwards compatibility.  Instead most dataframe operations would produce subclasses that had a well-defined common structure, as well as more custom attributes for their specific operation.  Let's look at a couple of examples.

```python
# API calls just create instances.  All logic happens there.
def read_parquet(file, columns, filters):
    return ReadParquet(file, columns, filters)

class ReadParquet(DataFrame):
    args = ["file", "column", "filters"]  # List of arguments to use when reconstructing
    inputs = []  # List of arguments that are DataFrame objects

    def __init__(self, file, columns, filters):
        self.file  = file
        self.columns = columns
        self.filters = filters

        self.divisions, self.meta = # do a bit of work on metadata
        
    def _generate_dask_layer(self) -> dict:
        ...

class ColumnProjection(DataFrame):
    args = ["dataframe", "columns"]
    inputs = ["dataframe"]

    def __init__(self, dataframe, columns):
        self.dataframe = dataframe
        self.columns = columns
        self._meta = self.dataframe._meta[columns]

    def _generate_dask_layer(self) -> dict:
        ...

class Add(DataFrame):
    args = ["left", "right"]

    def __init__(self, left, right):
        self.left = left
        self.right = right
        self.inputs = []
        if is_dask_collection(left):
            self.inputs.append("left")
        if is_dask_collection(right):
            self.inputs.append("right")

        self._meta = ...
        self._divisions = ...

    def _generate_dask_layer(self) -> dict:
        ...


```

As folks familiar with SymPy will recall, having attributes like args/inputs around makes it possible to re-generate a DataFrame automatically.  So if we do something like the following:

```python
df = dd.read_parquet(...)
z = df.x + df.y
```

Then this turns into an expression tree like the following:

```python
Add(
    ColumnProjection(
        ReadParquet(..., columns=None),
        "x",
    ),
    ColumnProjection(
        ReadParquet(..., columns=None),
        "y",
    ),
)
```

We can then traverse this tree (which is easy because we have a list of all attributes that are dask collections in the `inputs` attribute) and apply optimizaitons (which is easy because we can easily reconstruct layers because we have the `args` attribute).  

For example the `ColumnProjection` class may have an optimization method like the following:

```python
class ColumnProjection(DataFrame):
    ... # continued from above

    def _optimize(self) -> DataFrame:
        if isinstance(self.dataframe, ReadParquet):
            args = {arg: getattr(self.dataframe, arg) for arg in self.dataframe.args}
            args["columns"] = self.columns
            return ReadParquet(**args)._optimize()

        # here is another optimization, just to show variety
        if isinstance(self.datafarme, ColumnProjection):  # like df[["x", "y"]["x"]
            return ColumnProjection(self.dataframe.dataframe, self.columns)._optimize()

        # no known optimizations, optimize all inputs and then reconstruct (this would live in a superclass)
        args = []
        for arg in self.args:
            if arg in self.inputs:
                arg = getattr(self, arg)._optimize()
            else:
                arg = getattr(self, arg)
            args.append(arg)
        
        return type(self)(*args)
```

This is just one way of doing a traversal, using a method on a class.  We can do fancier things.  Mostly what I wanted to show here was that because we have args/inputs and class types it's fairly easy to encode optimizations and rewrite things.

What's notable here is that we aren't generating the graph, or even any semblance of the graph ahead of time.  At any point where we run code that requires something like `_meta` or `_divisions` from an input we stop all opportunities to change the graph under that stage.  This is OK for our internal code.  We can defer graph generation I think.  

I think that the main advantage to this approach is that we can easily reconstruct expressions given newly modified inputs.  I think that this is fundamentally what is lacking with our current HLG layer approach.  

However, this would also be a significant deviation from how Dask works today.  It's likely that this would affect all downstream projects that subclass Dask arrays/dataframes today (RAPIDS, Pint, yt, ...).  I think that that's probably ok.  Those groups will, I think, understand.

### Personal Thoughts

I've historically been against doing collection subclasses (the second approach), but after thinking about this for a while I think I'm now more in favor.  It seems like maybe we've arrived at the time when the benefits to doing this outweigh the complexity costs.  I think that this is motivated also by the amount of complexity that we're encountering walking down the HLG path.  

However, I do think that this will be hard for us to implement in an incremental way.  *If* we want to go down this path we probably need to think about an approach that lets the current `DataFrame` API persist for backwards compatibility with downstream projects (they wouldn't get any benefits of this system though) and a way where we can move over collection subclasses incrementally.

cc @rjzamora @jcrist

**Comment by rjzamora** [comment]
Thank you for taking the time to think through this problem and write this up.  I 100% agree that it will be a challenge to get what we really want out of the `Layer`-based approach.  As you stated: *“Today there is no real place where we capture a user's intent or their lines of code. We immediately create a task graph for the requested operation, dump it into a mapping, and create a new dask.dataframe.DataFrame or dask.array.Array instance”.*  As you know, this limitation has been driving me a bit crazy, because it feels like the “least-disruptive” solution is still very disruptive (likely requiring us to add a **lot** more complex machinery and structure to the existing HLG Layers).

For the reasons above, I am certainly open to the idea of making radical changes to the way Dask-collections operate.  To me, the only important consideration that you haven’t really discussed is: **How should collection subclasses interact with HighLevelGraph Layers (if at all)?** The general purpose of the HLG is currently two-fold: (1) To provide a “high-level” optimization mechanism, and (2) to avoid communicating a fully-materialized graph over the wire during distributed execution.  It seems like collection-subclasses can address both of these requirements.  However, we learned over the past year that enabling materialization on the scheduler can significantly limit the kind of logic that we can include in the corresponding `Layer` class.  Therefore, it seems reasonable that we will want `ReadParquet._generate_dask_layer` to produce a `Layer` object on the `client` (rather than a `dict`).

Since we will probably still need to deal with all the HLG materialization headaches with the new collection-subclass approach, I do want to establish that the Layer-focused approach would probably require about the same level of effort to achieve the same functionality. I say this, because we could still attach the same `inputs` and `args` information to the `Layer` class and regenerate a collection by traversing a tree in a very similar way. With that said, I do like the idea of separating the “state” of the collection from the HLG.  Also, I think I like the idea of using collection subclasses on the client, while continuing to use the HLG representation for communicating the graph to the scheduler.

**Comment by mrocklin** [comment]
If we had a good high level expression system on the collections side, what do you think the right design around layers would be?  Some of this stuff becomes redundant.

**Comment by rjzamora** [comment]
Right - exactly. It feels like the HLG becomes redundant unless we preserve a stripped-down version for serialization/materialization purposes only.  However, in the end the most-maintainable structure would probably remove HLGs altogether and simply require the same “clean” graph-materialization logic we already require in Layers.

**Comment by mrocklin** [comment]
The scheduler could grow a `def update_collection_graph` method or something similar that was a peer to the `update_graph_hlg` method.  I think that it's reasonable to think broadly here for a while.

**Comment by jsignell** [comment]
I really like the idea of using collections this way.  Historically have people proposed other approaches to this issue that were considered too high-level at the time?

**Comment by mrocklin** [comment]
People have certainly said things like "it would be great if Dask could do predicate pushdown", but I'm not sure that there has ever been a fully thought out proposal on expressions in Dask.  The efforts I've played with / proposed in the past have been for layers surrounding Dask.  That adds more complexity though.

**Comment by rjzamora** [comment]
>I think that it's reasonable to think broadly here for a while.

I totally agree - I am honestly very attracted to the idea of converting the collections themselves into something closer to symbolic expressions, and I like that you are thinking “radical.” In the previous two comments I was just making an honest attempt to play “devil’s advocate.”  I want to make sure we are acknowledging the fact that we will need to address similar materialization/serialization challenges to the ones we have faced with `Layers`.  Also, since we have already faced this for Layers, it is worth considering that the HLG class structure *could* also be redesigned as high-level expressions in a similar way.

Of course, I do expect that the long-term user API would be much more elegant with the collection approach.  It is actually somewhat similar to what NVTabular is currently doing (the “high-level” column-group dependencies and computations are defined separately from Dask).

**Comment by mrocklin** [comment]
I made the simplest possible system that I could think of in https://github.com/dask/dask/pull/7948.  It supports elementwise unary operations, binary operations, and  column selection.  It is also a valid dask collection and produces high level graphs. 

I don't like the way we use args/inputs, but these aren't important until we get optimizations around.  Other folks can probably suggest something smoother here.

**Comment by mrocklin** [comment]
So I think that the approach taken in that PR is somewhat smooth, in that it doesn't ask us to give up on HLGs as the communication mechanism to the scheduler.  We can shift things over smoothly without doing a massive disruptive change.  Then once we're ready we can change how we send things over.  

This doesn't fix any of the dask layer serialization stuff, but it also doesn't break it any further :)

In the meantime we get to solve things like your parquet metadata length problem.

**Comment by rjzamora** [comment]
>We can shift things over smoothly without doing a massive disruptive change

Nice!

>In the meantime we get to solve things like your parquet metadata length problem.

Woohoo!

**Comment by mrocklin** [comment]
I looked briefly at what it would take to implement reduction/apply_concat_apply and in general it seems like most of our methods are of the form ...

1. Scrape some metadata from the input dataframes (meta, divisions, ...) and make new meta and divisions
2. Do a ton of complex work to make an intesting graph layer
3. Wrap that into a HighLevelGraph
4. Wrap that into a new dataframe object

Really, we want to peel off layers 3 and 4, and just copy over step 1 into `__init__` and step 2 into `_generate_dask_layer`.  The main difference of the design is that these stages to all of our methods are broken up, and so we can easily redo `1` many times as we perform optimizations and such, and only do 2-4 when we're ready to compute.

**Comment by shoyer** [comment]
I think an expression system for dask arrays (and/or dataframes) sounds great!

One question is how closely tied this should be to Dask itself. I can see advantages (especially backwards compatibility) to using subclasses of existing dask data structures, but on the other hand I can see an abstract expression system for arrays potentially being useful for much more than only Dask.

**Comment by mrocklin** [comment]
I agree that a general abstract expression system for arrays would be
valuable.  If such a system existed then we would support its use.  I see a
path of modest cost to add this to Dask array/dataframe/bag and so barring
the existence of such a full solution that's probably what I would do.  I
do not personally intend to implement a full independent symbolic system
(at least not without a lot of support).

**Comment by mrocklin** [comment]
In https://github.com/dask/dask/pull/7948 I tried adding this sort of system directly into the existing class.  It's not great, but also not terrible.  It does allow for smooth changes and does allow us to reuse a lot of our code (I moved over the elemwise and aca functions, which account for a great deal of our methods).

**Comment by mrocklin** [comment]
OK, so taking a step back here we have a few classes with a few different requirements and considerations

### Collections

[Documented here](https://docs.dask.org/en/latest/custom-collections.html)

These require methods like `__dask_graph__`, `__dask_keys__`, and have methods like `__dask_optimize__`. In general these carry and manipulate dask graphs, which are mappings of individual tasks.

Subclasses of these typically also include attributes that show elementwise structure, namely `_meta`, as well as collection-wide structure, namely 

-  Bag: npartitions
-  Dataframe: divisions
-  Array: chunks

### Expr

These contain enough information to quickly recreate themselves (`args`) as well as information to traverse down to dependencies (`dependencies`).  Using these, as well as a class hierarchy, they could implement their own optimization methods.  It is important that little work is done on initialization, because we reserve the right to re-create these things frequently.  

## Can we do everything at once?

Sure!  But we need to be a little bit careful.  

In principle we could take the existing dask collection functions, turn them into classes which in the constructor do all of the following:

1. store the args/dependencies on the object (required by expr)
2. store the meta, divisions/chunks, and graph on the object (required by collections)

Step 2 here is what is already written, so adding this into the constructor would be the easiest path, but this can also be slow.  Instead we want to move the slow parts of 2 to some other function (currently named `_generate_dask_layer`) but we want to keep the structural parts of 2 (meta, divisions/chunks) in the constructor like the following:

1.  **On initialization:** Create args/dependencies for expr and meta/divisions for collection
2.  **On demand later on:** Create graph

The challenge here is that some of our code creates the divisions and the graph and the meta at the same time.  We need to tease apart this creation.  I suspect that this is going to be possible almost all of the time, and in the other times we can always just create and store the graph in the constructor (slow, but not the end of the world).  Really, we just need to go through and tease apart all of our methods.  I think that calling out these necessary pieces of metadata is useful.

**Comment by mrocklin** [comment]
For args/dependencies I've gone through a few different options of representation.  

As a reminder we need these for the following reasons:

1.  **args**: we need to be able to recreate an object with different inputs.  Something like the following should always work

    ```python
    new = type(obj)(*obj.args)
    assert new == obj
    ```

2.  **dependencies**: we need to be able to see the other dask expressions on which this one was based so that we can traverse down this tree when optimizing

### How do we store these things?

I first started with the names of attributes, like the following:

```python
class BinOp:
    args = ("operator", "left", "right")
    dependencies = ("left", "right")

    def __init__(self, operator, left, right):
        ...
```

(we should traverse down left and right, but don't need to traverse down the operator, which is unlikely to be a dask expression.

However this is challenging because there are a lot of getattr calls, and often the dependencies at least aren't known ahead of time.  For example in the case of `x + 1`, where expr.right is not a dask collection.

So in the more recent implementation when I change the dd._Frame class directly I use attributes and properties 

```python
class Binop:
    def __init__(self, operator, left, right):
        ...

    @property
    def args(self):
        return self.operator, self.left, self.right

    @property
    def dependencies(self):
        out = []
        if is_dask_collection(self.left):
            out.append(self.left)
        if is_dask_colleciton(self.right):
            out.append(self.right)
        return out
```

This seemed to work better.

I also found that using only `*args` was limiting, because in many cases we encourage the use of **kwargs.  I wonder if instead we want the invariant `new = type(obj)(*obj.args)` to instead be `new = type(obj)(**obj.parameters)`, where we use a dictionary with named parameters instead.  This seems more predictable.

**Comment by mrocklin** [comment]
In dask dataframe I ran into problems with subclassing.  Currently we have four classes, DataFrame, Series, Index, and Scalar.  (More if you include RAPIDS and GeoPandas).  I found that I had to make classes like `ElemwiseDataFrame`, `ElemwiseSeries`, and so on.  These were very simple definitions

```python
class ElemwiseDataFrame(ElemwiseBase, DataFrame):
    pass
```

But still it was a code smell that pointed to some of the bad behaviors of object inheritance.  I suspect that this approach would put us in even more pain when it came to downstream projects like RAPIDS/GeoPandas, and also maybe Pint.

A common approach in this case is to switch from inheritance to encapsulation.  I don't yet have a clean design for that.

**Comment by mrocklin** [comment]
@quasiben @rjzamora and I spoke this morning (along with some other folks).  It looks like @rjzamora has time and interest to look into executing something like this starting around one-to-two weeks from now.  I would like for folks here to come to some consensus on a path forward before then.

**Comment by douglasdavis** [comment]
This has been fun to read (along with https://github.com/dask/dask/pull/7948/). I'm really interested in helping in the effort. w.r.t. consensus on a path forward: do you mean having an answer to the handful of proposed questions that you've laid out in the issue so far? Like property vs. attribute args/dependencies, `*args` vs. `**parameters`, and inheritance vs encapsulation? Or something more broad, like incremental implementation vs all at once? Or all of the above!

**Comment by mrocklin** [comment]
> consensus on a path forward: do you mean having an answer to the handful of proposed questions that you've laid out in the issue so far

Mostly I'm looking to hear from people who have different experiences than me and who can anticipate issues that I haven't.  My expectation is that this will mostly come from some of the other grizzled maintainers :)

**Comment by mrocklin** [comment]
> I'm really interested in helping in the effort.

I'm glad to hear this.  My sense is that when we first start there will lots of false starts and redesigns that will be frustrating for anyone who is not familiar with and comfortable with tearing apart Dask's design.  (this is what I'm going through now).  However, I think that once we find a path that seems to be working then there will be a lot of work that I think many people could probably participate in in parallel more effectively.  

In particular, I expect @rjzamora to focus his work on Dask Dataframe (due to the RAPIDS connection) but probably not invest as much effort in Dask array.  Doing the rewrite process within Dask array would probably be a good month-long project for someone if they were interested.

**Comment by dhirschfeld** [comment]
> I agree that a general abstract expression system for arrays would be valuable. 

That sounds a bit like what [`metadsl`](https://github.com/Quansight-Labs/metadsl) was intended to support? Unfortunately development has stalled there but think the concept has a lot of merit so I'm very interested to see this proposal evolve...

**Comment by douglasdavis** [comment]
> In particular, I expect @rjzamora to focus his work on Dask Dataframe (due to the RAPIDS connection) but probably not invest as much effort in Dask array.

Yea I've sensed that this is the case

> Doing the rewrite process within Dask array would probably be a good month-long project for someone if they were interested.

Happy to target this. Will keep an eye on the path planning

**Comment by crusaderky** [comment]
I think that this design would be incompatible with ``dask.graph_manipulation`` and any other (third party?) code that perform collection-agnostic graph alterations. The way that module works is:

1. fetch the HLG from the collection with ``__dask_graph__()``
2. add its own layers on top
3. rebuild the collection using ``__dask_postpersist__(..., rename={old_collection_name: new_collection_name})``)

The module has no clue about which specific collections it's working with.
Without HLG, it would be forced to completely materialise the graph client side - or am I missing something?

**Comment by mrocklin** [comment]
I gave encapsulation some thought today.  I didn't actually do any work on it though, so these thoughts may be completely silly once we hit actual code. 

There is an open question of where all of the necessary metadata sits.  This metadata includes task graph construction, meta object, divisions, expression tree, and the dataframe/series/index API type.  

The challenge of subclassing is that we want the API to depend strongly on the dataframe/series/index API type, so that needs to have a strong influence on the class.  Adding the expression tree into that typing structure creates two different dimensions and so an cartesian product of subclasses.  

So we leave the top-level class to the API type (dataframe/series/index) and have that contain an expression

```python
class FrameExpr:
    ...

class DataFrame:
    expr: FrameExpr
    ...
```

While all of the graph, metadata, divisions, etc need to be accessible from the DataFrame/Series/Index classes, it's not clear where the logic for these things should live.  Should they live on the DataFrame/Series/Index classes (I'm going to start calling these user API classes) or should they live on the expression classes (Elemwise/IO/ReadParquet/Reduction/...)?  This is an open quesiton.

My gut reaction right now though is that *everything* should go on the FrameExpr.  This leaves the user API classes to be quite  simple

```python
class DataFrame:
    def __init__(self, expr):
        self.expr = expr

    @property
    def _meta(self):
        return self.expr._meta

    ... # and so on
```

This ends up being a bit less clean from an implementation perspective.  We need to change *all* calls to functions like `new_dd_object`, which is unpleasant.

## Backwards compatibility

Or not?

We should have a fallback, an expression that contains a task graph/meta/divisions without provenance.  

```python
class LeafFrame(FrameExpr):
    def __init__(self, dsk, name, meta, divisions):
        ...
    @property
    def args(self):
        return ()

    @property
    def dependencies(self):
        return ()
```

And then current API points that create dask dataframes, like `new_dd_object` will have a legacy mode

```python
def new_dd_object(dsk=None, name=None, meta=None, divisions=None, expr=None):
    if expr is None:
        expr = LeafFrame(dsk, name, meta, divisions)
    return get_parallel_type(expr._meta)(expr)
```

I don't think that calling `dd.DataFrame` is particularly common any more, but we can probably do the same thing there.


## Adding actual operations

Functions like elemwise will move their logic into classes, as before, and then they'll create a new_dd_object with that expression

```python
def elemwise(*args, **kwargs):
    expr = Elemwise(*args, **kwargs)
    return new_dd_object(expr)
```

**Comment by asmeurer** [comment]
> Mostly I'm looking to hear from people who have different experiences than me and who can anticipate issues that I haven't. My expectation is that this will mostly come from some of the other grizzled maintainers :)

Here are some thoughts from a SymPy maintainer. You'll have to pardon my ignorance of the dask internals. If something I say here doesn't make sense, please let me know.

- My first thought when I saw this idea is that the expression system would be separate from the classes. However, Matt makes a good point that this may require keeping the separate things in sync too much. You considered making the collections classes themselves the expression tree classes, and [having collections link to separate expression tree classes](https://github.com/dask/dask/issues/7933#issuecomment-892087327), but what about the reverse, expression tree classes that contain the collections? Anyway, it's not clear to me yet which is the best option. I honestly don't feel I have enough knowledge of the dask internals to say. 

- My biggest question here is whether or not these collection objects you plan to make into expression trees are immutable. If they are, you should be good to go. If they aren't, you may run into a lot of headaches. The issue being that if you mutate an expression that may require updating any tree that contains it, but expression tree DAGs point down the tree not up (expressions don't know which other expressions contain them).

- Regarding using `type(*obj)(**obj.parameters)` instead of `type(obj)(*obj.args)`, that sounds good. The only limitation of `**parameters` is that to support an actual `*args` signature you'd to also support passing them in as a parameter. But that's should be easily doable with `def func(*args, **, _args=())` or something. You also wouldn't be able to use Python 3.8+ positional-only args.

I will say that one thing I have always found nice about the SymPy expression system that I would recommend emulating is that every expression tree can be uniformly walked using `.args`. Compare this to the Python ast module, which only has attributes, and a list of the attributes. The resulting system is annoying to work with by hand, because you have to know the exact type of each node in order to walk down it. You are effectively forced to use the visitor pattern to examine and manipulate AST objects. But the visitor pattern is only one way to manipulate expressions. It works well when you have different code for each expression tree type, or when you want the visiting operations to be extensible. But it is also quite verbose, as it requires you to write a class for even the simplest manipulation. There are other ways, which are more terse and simpler to reason about. For example, you could just use recursive functions that walk through the tree. Another way, which I know Matt is fond of (because he wrote it), is a functional API like that found in [sympy.strategies](https://github.com/sympy/sympy/tree/master/sympy/strategies), where you have a lot of expression manipulation functions that compose nicely, allowing you to write code like `bottom_up(f, expr)` (see the [sympy.simplify.fu](https://github.com/sympy/sympy/blob/master/sympy/simplify/fu.py) module for a large example of code written in this style). Yet another way is based on pattern matching, e.g., using something like MatchPy (which is indicendally the route I would recommend for doing expression optimizations). The point isn't that one way is better than any other. All methods have their valid uses. The point is that the simpler expression tree design doesn't lock you into only one way of thinking about manipulation.

SymPy has the additional restriction that every element of `.args` should itself be a SymPy class, which has both advantages and disadvantages (I know Matt and I have disagreed about that in the past). The main disadvantage of not making this requirement is that you always have to be on guard for non-expression tree classes when walking an expression tree. It especially hinders the second type of tree walking I mentioned above (manual recursive functions). On the other hand, the main disadvantage of having the requirement is that you have to have classes to wrap basic collection types like tuples or str. And for instance the BinOp class in https://github.com/dask/dask/pull/7948 wouldn't be possible in SymPy. Each operation has to have a separate class (not that that's necessarily a bad thing; it's not clear to me from a pattern matching perspective whether BinOp would be a good idea). Matt's idea of separating `args` and `dependencies` might be a clean way of dealing with this issue, assuming I understood correctly the purpose of that distinction (although I would personally change `dependencies` to just `deps` or something else that is shorter to type). 

I do think you can make the design here simpler than SymPy's in many ways. SymPy uses `expr.func(*expr.args)` instead of `type(expr)(*expr.args)` to allow for expressions' head object to not necessarily be its class. But for a simple expression tree design, this is unnecessary complexity. As another point, you should avoid the temptation to do automatic simplification on the expressions at instantiation time, instead preferring to represent exactly what the user input. All simplification, even "trivial" simplifications, should be done separately as optimizations (although any kind of type checking that would lead to an error can be done in the constructors). In SymPy, basic simplifications like `x + x -> 2*x` happen, but what this means internally is that `Add(x, x)` has to return `Mul(2, x)`, so the structure requires the use of `__new__`. Automatic simplification is also a bane for many other reasons. See https://github.com/sympy/sympy/wiki/Automatic-Simplification for a list of some of the problems it can lead to.

**Comment by mrocklin** [comment]
Thank you for the comments @asmeurer .  Some responses! 

-  So far I like expressions being in collections.  This keeps the user API the same as it is today.
- I think that mutability will be ok.  Expressions can be immutable, but operations on a collection can create and swap out a new expression.  We do this today with Dask graphs.  It is also what Theano did back in the day.
- "every expression tree can be uniformly walked using .args" Yup.  Totally agreed.  This is definitely a part of SymPy that I would like for us to steal.
- Regarding traversal techniques I think that we can leave this for a future stage.  Just getting expressions, and doing a dead simple optimization routine (maybe just with recursive method calls) seems like a good first step.  We can do lots of things after we have that base.
- I like the args/deps separation.  I think that it is a pragmatic choice.  It's been a while, but I think that we might have done this in Blaze?  (my memory is hazy.)  If so then I think it might have worked out well?  
- I agree with avoiding automatic simplification.  I'd like to avoid the use of `__new__` (I suspect that you will agree)

**Comment by asmeurer** [comment]
> I think that mutability will be ok. Expressions can be immutable, but operations on a collection can create and swap out a new expression. We do this today with Dask graphs. It is also what Theano did back in the day.

It sounds like you are saying that the design of collections containing corresponding expressions makes it so that collections can be mutable with expressions immutable. If collections are indeed mutable, I agree this is the best design.

> Regarding traversal techniques I think that we can leave this for a future stage. Just getting expressions, and doing a dead simple optimization routine (maybe just with recursive method calls) seems like a good first step. We can do lots of things after we have that base.

This sounds good. It might be worth getting some basic familiarity with MatchPy to make sure you don't design your way out of the possibility of using it. But other than that, my point is that as long as you have a simple expression tree structure, different kinds of traversal techniques are straightforward to implement.

> I think that it is a pragmatic choice. It's been a while, but I think that we might have done this in Blaze? (my memory is hazy.) If so then I think it might have worked out well?

I never worked with Blaze. I'll have to take a look at how it was designed.

**Comment by mrocklin** [comment]
> This sounds good. It might be worth getting some basic familiarity with MatchPy to make sure you don't design your way out of the possibility of using it. But other than that, my point is that as long as you have a simple expression tree structure, different kinds of traversal techniques are straightforward to implement.

We'll have to make sure that someone who works on this is familiar with MatchPy ;) 

> I never worked with Blaze. I'll have to take a look at how it was designed.

I wouldn't look too closely :)

**Comment by wesm** [comment]
Hopefully this isn't too OT, but I wanted to make Dask folks aware that we have an adjacent effort/discussion under way in Apache Arrow to develop a language-independent serialized expression format connecting user SDKs (e.g. things like Ibis or even dask.dataframe potentially) with Arrow-based/Arrow-compatible computing engines:

https://lists.apache.org/thread.html/r90c1e6ba1c7b27a960df3f27dc2c1aeb542f07bb7af72c17138667f0%40%3Cdev.arrow.apache.org%3E

The idea is to provide a common ground consisting a standard lower-level operators rather than each engine having their own slightly idiosyncratic front end (e.g. using a certain SQL dialect, or something else) — there would be the presumption of using Arrow's schemas / type metadata to describe input/output types of operations.

In the same way that Dask uses pandas as a dataframe computing engine, it would be interesting to see if this High Level Expression effort could lead to a world of pluggable single-node engines for Dask to orchestrate (e.g. using one of the various Arrow-based engines — Rust, C++ — in development or even an small-footprint embeddable analytic SQL engine like DuckDB — which can now input/output Arrow easily). One might imagine that dask.dataframe in the future could not use pandas at all and permit use of alternative (or even multiple) backends to execute various tasks (and if useful, Arrow could be used as a serialization middleware).

**Comment by mrocklin** [comment]
Thanks for jumping in @wesm .  That seems like an important topic to me.  I've made a dedicated issue for it at https://github.com/dask/dask/issues/8074

For the purposes of this conversation I think that Dask will need a high level expression system that is separate from Arrow's compute engine expression serialization format.  I don't think that this is what you were suggesting, but I doubt that Arrow would find the things that Dask needs from such an expression system in-scope.

**Comment by mrocklin** [comment]
Quick logistical update.  The plan here was for Rick or I to take a first pass at implementing the plan above before Aaron came back from vacation.  However, Rick was sidelined by Parquet issues and I was sidelined by shuffling ones.  As a result we're a little bit behind schedule.  

@asmeurer I don't know if you are comfortable enough to try this blind (I imagine that you'll run into a couple of false starts, which is fine by me) of if you want to wait for someone else to take the first step.

**Comment by asmeurer** [comment]
How long do you think it would take to do whatever it was you were planning on doing? I have other projects I can work on in the interim, but I'd say I'd probably like to get started no later than next week, so if it would take longer than that, I should probably just try to jump in.

**Comment by mrocklin** [comment]
I'm fully supportive of you just jumping in and was hoping that this is
what you were going to propose.  I like this just because it gets us moving
faster.  It does increase your frustration load I think though.  I suggest
that you take on the dev burden here and that you and I check in frequently
this week if that works for you.  I'll follow-up off-list as well.

**Comment by mrocklin** [comment]
Just to reiterate a starting place I suggest starting with https://github.com/dask/dask/issues/7933#issuecomment-892087327 . This might proceed in a few steps:

## Create DataFrameExpr class, support backwards compatibility

We would make a DataFrameExpr and a DataFrameLeafExpr class (better names welcome).  

We would then modifying the existing DataFrame/Series/_Frame classes to consume mostly just that expr (as mentioned in the comment above) and change the new_dd_object function to support the old API, but create LeafExprs internally and then hand them to the appropriate API class (dataframe, series, index, ...).

## Create non-trivial expressions

Now that everything works but doesn't provide any benefit, we might want to try changing over some of the more base functions.  The biggest one here is elemwise, which is used very often throughout the codebase.  

After this would probably be apply_concat_apply, what we use for reductions.  

After this would probably be some IO thing, like read_parquet.

## Circle back

At this point we've probably learned a lot and we'll want to revisit the design

**Comment by wesm** [comment]
> I think that Dask will need a high level expression system that is separate from Arrow's compute engine expression serialization format. I don't think that this is what you were suggesting

Indeed it wasn't what I was suggesting, so wanted to reiterate that I see these are parallel/complementary efforts, but where each can learn from each other (e.g. we want to make sure that the Arrow serialized expressions can support a lot of the concepts that a dask.dataframe-like interface would produce)

**Comment by asmeurer** [comment]
What is the best way to compare two meta objects to see if they are equal?

**Comment by rjzamora** [comment]
>What is the best way to compare two meta objects to see if they are equal?

You could probably start with [assert_eq](https://github.com/dask/dask/blob/ab69fe124eb36a63c092026ab082ccb08311fc79/dask/dataframe/utils.py#L525) for now.  We could probably work out a more-streamlined check if one doesn't already exist.

**Comment by asmeurer** [comment]
I pushed up my very initial thoughts at https://github.com/dask/dask/pull/8084. I'm probably going to play around with this some more tomorrow, but if you have any initial thoughts on what I am doing before then, I'd love to hear them.

**Comment by jcrist** [comment]
Ok, I've finally had a day to poke deeply at this. I don't have any real code
to show for it (lots of scrapped implementations), but I've come up with a few
design issues/questions we'll want to tackle.

## 1. Structure of the expression datastructure

There are a few ways to structure an expression datastructure.

**Direct references**

All of the discussion above represents it as a DAG of `Expr` node objects, each
of which may contain zero or more direct references to child nodes (as well as
any extra metadata needed).

For example:

```python
from collections import namedtuple

Add = namedtuple("Add", "name, x, y")
Const = namedtuple("Const", "name, value")

a = Const("a", 1)
b = Const("b", 2)
c = Add("c", a, b)
d = Add("d", b, c)
```

This is nice because any recursive dependency of a node must be a recursive
child reference of that node. This means that if a rewrite pass rearranges a
few nodes in the DAG (or drops a node due to an optimization), the total
expression remains self consistent since the state is always localized
(there aren't extra references we'd have to keep up-to-date/cull).

The downside is that since computations are a DAG not a tree, there can be
multiple references to the same node. Any graph traversal (optimization pass,
low-level-graph generation, ...) would need to be memoized to avoid doing
repeated work. Additionally, any rewrite of a subexpression in the graph would
have to traverse the whole expression to find and substitute out the
subexpression.

**Indirect references**

A different way of writing the same expression system would be to "flatten" the
structure and use *indirect references*. Nodes no longer contain direct
references to other nodes, instead they hold a reference to the node names. For
example:

```python
from collections import namedtuple

Add = namedtuple("Add", "x, y")
Const = namedtuple("Const", "value")
ExprDag = namedtuple("ExprDag", "exprs")

ExprDag(
    dict(
        a=Const(1),
        b=Const(2),
        c=Add("a", "b"),
        d=Add("b", "c"),
    )
)
```

This representation means that rewriting a single subexpression requires no
traversal, since all references are done indirectly through the name->expr
dictionary. 

This also makes it easier to merge the expressions for multiple objects during
a `compute` call. For example, say we call:

```python
dask.compute(c, d)
```

on the expressions above. With the direct-reference representation we'd need to
create a new node (e.g. `TupleExpr(c, d)`) to merge the expression dags
together, then optimize the whole structure. With the indirect references, this
is just creating a new `ExprDag` object with the `.exprs` dicts merged.

The downside here is that any rewrite that drops a node would
result in dead nodes still remaining in the structure (a pass to cull these
would be required).

---

The second (indirect references) representation is what we do for the existing
`HighLevelGraph` objects, except the expressions are called "layers".

After thinking about the pros/cons of each approach, and the types of
operations we'd want to do on the expression system, I'm mildly leaning towards
the second approach.  I think it will be easier to work with for the operations we want
to do (cc @asmeurer, you may have thoughts on this).

But if we go with that approach, why not just modify the existing `HighLevelGraph` layers
to contain the extra metadata we need (since the overall datastructure is basically the same)?

## 2. How do HighLevelGraphs fit into this bright new future?

We already tried an expression system with the `HighLevelGraph`/`Layer`
objects. These were created to do two things:

- Reduce I/O when sending large graphs from the client -> scheduler, since we
  could send some layers symbolically rather than fully materializing the graph
  on the client side.

- Provide a place to perform some higher level optimizations.

Whatever approach we use should be able to handle all the optimizations that
the existing `HighLevelGraph` system can handle (since if you squint, they're
basically the same but with more metadata attached), so there's no reason IMO
to keep optimizations at that level. So the question is, do we still use
`HighLevelGraph` layers to reduce IO to the scheduler, or do we directly send
our expression objects instead?

I'm going to argue that for the purpose of reducing I/O, we probably need fewer
types of "Layer" to send between the client and scheduler than we'd need types
of expressions to perform optimizations. This may be clearer with an example.

For example, say for `dask.bag` we want to rewrite `.count()` operations before
`.map()` operations, since `bag.map(func1).map(func2).count()` is equivalent to
`bag.count()`. To do this, we'd need a way to distinguish `.count()` nodes from
other expressions in the tree (perhaps with a `BagCount` expression, idk).
However, the scheduler doesn't actually care about this information, the
scheduler only cares about the low-level graph. To the scheduler, symbolic
layers are only a cheaper way to communicate the graph structure. So the number
of graph layer representations the scheduler would need to understand is much
lower than the number of expressions the client would need for optimizations.
Separating these representations would make it easier to have mismatches in
Dask versions between the client and scheduler. It also makes it easier for
external libraries to create their own expression nodes, since the expression
classes would only have to exist client side (provided they could be lowered
into one of the representations the scheduler understands).

## 3. A suggested path forward

After poking at this for a bit, I think this is less of an engineering design
question, and more of an engineering process question. How can we convert an
existing large ungainly system like the existing `HighLevelGraph`
representation into something that looks kinda similar but contains more info,
without breaking everything or preventing biweekly releases?

I see two paths forward:

- Iterate on the existing `HighLevelGraph` system to add additional layers/info
  to layers to support the optimizations we want to do. In the process we'd
  likely want to fix-up many of the warts that this system has (for example, I
  find the serilization mechanism here quite confusing). This is likely to be
  painful as the existing code paths are large, but *may* be easier to do
  iteratively. Not sure.

- Create a brand new expression system that looks like the `HighLevelGraph`
  representation from scratch, but does it "right". This representation would
  lower to a few existing `Layer` classes (e.g. `Blockwise`) for serialization
  alone, but we'd want all optimizations to be handled by the new expression
  system. Once that's in place, we'd then start to cull back the now extraneous
  layers, and cleanup the `HighLevelGraph`/`Layer` objects. This has all the
  joys and pains of writing new code.

Neither process sounds that enjoyable, since there's a lot of existing code
that will have to be modified.

I'm going to punt on answering the path-forward question and ask those that
have been more involved in the previous HighLevelGraph work to chime in here
with some thoughts (cc @gjoseph92, @rjzamora, @jrbourbeau,
apologies if there are others I should ping, not sure who all was involved).

Either way, I think the compute pipeline we end up with is:

- User calls `dask.compute(...)`
- The expressions for all objects are merged into some larger expression so
  common sub-expressions can be shared
- Optimization passes are applied to this expression, resulting in a new
  simpler/more efficient expression.
- This expression is then lowered into a serializable graph representation to
  send to the scheduler. Ideally no optimizations are performed at this level.
- The scheduler receives these layers, and materializes any that were sent as
  symbolic representations.

**Comment by mrocklin** [comment]
Thank you for the writeup @jcrist .  I've only skimmed it so far, and plan to go through it more thoroughly tomorrow.  However, some quick notes:

I'd like to add a third approach to how to handle dependencies

1.  Use a tree, but be careful about duplicated expressions
2. Use a named layer approach
3. (new!) Use a `.dependencies/children` attribute as in the current plan, but also add a `.dependents/parents` attribute that goes the other direction.  This would be a Weakset to avoid a reference mess.  Optimizations would then look both at children and parents when deciding what to do

I'm not pushing for one over the other, just adding another option for consideration.

I like the idea of reusing/building on HLGs if we can "do them right".

**Comment by asmeurer** [comment]
I think my homework now is to take a closer look at HighLevelGraph, so I can't really say much here until I do that. It does seem to me from what I've done so far that 1) we don't really want expressions to be tied to closely to the container classes and 2) the expression system is very similar to the graph, just with more higher level information. So it might make sense to merge them. OTOH, the expressions being higher level could in fact mean we *don't* want them to be the same, since that would make it harder to separate high level and low level details. So I think the question boils down to just what it is the high level graph is doing—and isn't doing—compared to the low-level graph. 

I will say this about your direct vs. indirect references approaches. The indirect references approach makes walking the expression tree a bit more complicated, since you always have to dereference node names. This is basically the same approach used by the dask graph itself, so you might have a better sense for how complicated this gets than I would. I'm actually curious how the general graph code in dask handles repeated references. Is the logic for this complicated? Does using named references simplify avoiding repeated work on duplicated noes? Is there some part of dask I should look at to get a better idea of this myself?

I would reiterate my comment https://github.com/dask/dask/issues/7933#issuecomment-895552688 that having a uniform way of walking the expression tree should not be underrated, especially for things like optimization passes which amount to tree rewriting. I'm still quite partial to the SymPy `.args` approach and would lobby pretty hard for it.

I expect some of your downsides to direct references can be mitigated without too much difficulty. For instance, one could create a flag dag graph from a direct reference graph to hold alongside the expression. This would basically amount to memoization, but storing the info on repeated nodes on the top-level expression rather than on the function. Or another option could be to use reverse references as @mrocklin suggested.

**Comment by asmeurer** [comment]
Question: The documentation at https://docs.dask.org/en/latest/high-level-graphs.html says "For example, in the case above we may want to automatically rewrite our code to filter our datasets before adding 100:" Is this optimization implemented somewhere? I saw optimizations.py but I couldn't tell if any of the functions in it actually do this. This is obviously the sort of optimization we want to do. If it doesn't exist yet, I it would be useful to think about what it would look like.

**Comment by jcrist** [comment]
> (new!) Use a .dependencies/children attribute as in the current plan, but also add a .dependents/parents attribute that goes the other direction. This would be a Weakset to avoid a reference mess. Optimizations would then look both at children and parents when deciding what to do.

Since this information is easy to compute later (and once computed could be maintained as part of the optimization passes, as we do now), I'm skeptical of the benefits of maintaining bidirectional pointers. Usually you want to find nodes that have only one dependency/dependent (for e.g. fusion) or all dependents of a certain node. Having an external mapping computed during the optimization pass that tracks this info doesn't seem too onerous.

> I'm actually curious how the general graph code in dask handles repeated references. Is the logic for this complicated? Does using named references simplify avoiding repeated work on duplicated noes?

During most optimization passes we compute the dependencies of all tasks in bulk at the start of the pass (as a mapping from task name to dependencies). If needed, this mapping can be inverted to find dependents. Some passes are able to update the dependencies structure dynamically and forward this on to subsequent passes to avoid repeated work. It's a bit of a bookkeeping mess sometimes, but it's also worked well enough so far.

**Comment by jcrist** [comment]
I'd really like to hear from some of the people that worked on the HighLevelGraph stuff. Rebumping:

> I'm going to punt on answering the path-forward question and ask those that
have been more involved in the previous HighLevelGraph work to chime in here
with some thoughts (cc @gjoseph92, @rjzamora, @jrbourbeau,
apologies if there are others I should ping, not sure who all was involved).

**Comment by jcrist** [comment]
Question: if during an optimization pass we change the inputs to a call such that they would have hashed differently, do we regenerate the layer name? Or do we leave it as is? For example:

```python
df1 = dd.read_parquest("data/*.parquet")  # This currently hashes based on only the file paths
res1 = df1.x + df1.y  # with column projection, we can rewrite the above to only read `x`, `y`

df2 = dd.read_parquet("data/*.parquet", columns=["x", "y"])
res2 = df2.x + df2.y
```

Here we have `res1` and `res2`, which should have the same result. The first relies on column projection to automatically cull the loaded columns. The second specifies it manually. Do these share structure in the low level graph? If we hash during expression generation (and thus "freeze" the node names), even after the rewrite pass that makes them equivalent they'll have different names. BUT making names dynamic would mean that some nodes might have a different name everytime `.compute()` is called (e.g. different expressions computed on `df1` may result in different column subselections needed each time).

Right now I'm thinking that we don't handle this, and names are frozen at creation time, based only on the parameters passed by the user. This would prevent some optimizations (e.g. we couldn't treat `res1` and `res2` as the same), but I'd argue that those cases are likely rare and contrived. It seems easier to make names frozen on creation time.

**Comment by asmeurer** [comment]
I think the complexity here is thinking about how we handle duplicate nodes in a sane way. In other words, we really do have to think of expressions as a DAG, not just a tree. This will probably require some adaptation on my part. I'm used to SymPy. In SymPy, if you have an expression like `(x + 1)**2 + (x + 1)**3`, the `x + 1` is just repeated twice in the expression tree. The literal Add class may or may not be de-duplicated in memory due to caching/memoization, but that's an implementation detail which ultimately doesn't matter because 1) expressions are immutable, and 2) == always compares expression by exact symbolic equality. But in dask, if you have `df = read_csv('f.csv') + 1; df[df.name == 'Alice']`, you don't want to represent that as `(read_csv('f.csv') + 1)[(read_csv('f.csv') + 1).name == 'Alice']`. Functionally they are the same, but computationally they aren't. 

Hence dask is built with graphs rather than expressions. The graph lets you explicitly compute `read_csv('f.csv')` exactly once and reuse the result. Ultimately what we are building here is an optimizing compiler. 

My first question about this is whether we can always assume a computation is functionally sane like this. Obviously if you start adding mutations in the mix computations can't be reordered so simply, but does dask even allow that sort of thing in the first place? The [docs](https://docs.dask.org/en/latest/graphs.html#don-t-modify-data-in-place) say that side-effects are not recommended, but to what degree can we assume there aren't any?

**Comment by jcrist** [comment]
>My first question about this is whether we can always assume a computation is functionally sane like this. Obviously if you start adding mutations in the mix computations can't be reordered so simply, but does dask even allow that sort of thing in the first place? The docs say that side-effects are not recommended, but to what degree can we assume there aren't any?

We let users write some things using the high-level api that _look_ like mutation. But no low-level task (the thing that is actually in the low-level graph) should ever mutate one of their inputs. So while you can do:

```
df = make_some_dask_dataframe()
df["a new column"] = df.x + df.y
```

The actual pandas objects during execution are never mutated. Instead a copy is made, a new column added, and the result returned.

**Comment by asmeurer** [comment]
> Here we have res1 and res2, which should have the same result. The first relies on column projection to automatically cull the loaded columns. The second specifies it manually. Do these share structure in the low level graph? If we hash during expression generation (and thus "freeze" the node names), even after the rewrite pass that makes them equivalent they'll have different names. BUT making names dynamic would mean that some nodes might have a different name everytime .compute() is called (e.g. different expressions computed on df1 may result in different column subselections needed each time).

I'm probably missing some context here. Why would they have different hashes? Does the hash include some sort of timestamp that makes the exact same expression unique?

I guess I'm also not really understanding why this matters. Do you foresee this being a problem during optimization passes, perhaps related to the "duplicate computation" problem I talked about in my previous comment? It seems to me that if these collapse, it would only be by coincidence. What if it instead optimized it to `df2 = dd.read_parquet("data/*.parquet", columns=["y", "x"])`? Then they would be equivalent but would not look the same. Effectively, node de-duplication would need to be its own optimization pass, which either applies some canonicalization to the nodes so they tend to look the same if they are the same, or does some sort of looser equality. In general, such things would necessarily be heuristic and there would always be cases that it would miss (the general problem would almost certainly be Turing complete). 

However, if hashes really do contain more than just the exact input, then I think that means we shouldn't use it for equality (as Jim suggested to me on a recent call), because you can have situations where two expression should be exactly equal simply because they were copied in the tree. Just take the `(read_csv('f.csv') + 1)[(read_csv('f.csv') + 1).name == 'Alice']` example I talked about above). If `read_csv('f.csv') + 1` isn't guaranteed to hash to the same thing both times, we would need an `__eq__` method based on something other than hash that compares them equally. To be sure, I don't think is a particularly hard problem if this is the case. It's just something I wanted to mention since it relates to this.

**Comment by jcrist** [comment]
Currently all dask collection objects (e.g. `dd.DataFrame`) have a `name` that usually becomes part of the keys for that operation in the low-level graph (e.g. `(df.name, 0)` would be the key for the 0th partition in a dask dataframe). This name is usually computed by intelligent hashing of the arguments to the method that produced that object (the `tokenize` function is what handles this). For example, for `random_split` the name is composed of a hash of the input frame, `frac` and `random_state`: https://github.com/dask/dask/blob/9f0b00f039a7775dccccffe59d6802468508d9f7/dask/dataframe/core.py#L1047-L1049. I say we "intelligently" hash, since operations are free to canonicalize their arguments in any way before calling `tokenize` to handle things like order-independent operations, etc... This system has served us well, and provided a very simple way of pulling out common subexpressions to avoid duplicated work.

When moving to an expression system, we won't be eagerly generating the low-level-graph anymore, which lets us  do more interesting rewrite passes. However, we'll still need to make `df.name` available as part of the API for any custom algorithms that won't be representable as part of the expression system.

If we eagerly generate the name before applying optimization passes, then two nodes that would be the same post-optimization won't hash the same pre-optimization. The column projection example I gave above is one case of this.

**Comment by asmeurer** [comment]
> If we eagerly generate the name before applying optimization passes, then two nodes that would be the same post-optimization may not hash the same pre-optimization. 

Isn't that expected? In your example, they *aren't* the same. `res3 = df1.x + df1.y + df1.z` would work but `res3 = df2.x + df2.y + df2.z` would not. Requiring df1 and df2 to know this requires global knowledge of the entire graph, which complicates things. 

> I say we "intelligently" hash, since operations are free to canonicalize their arguments in any way before calling tokenize to handle things like order-independent operations, etc... This system has served us well, and provided a very simple way of pulling out common subexpressions to avoid duplicated work.

This worries me, regarding using hash for equality. Canonicalization at construction should be avoided. You can do anything that you would be done with canonicalization as an extra pass on the expression, but the problem is that you can't undo canonicalization. And it becomes hard to remove once it is there. I should note that very things like kwargs being undered would happen anyway if you used frozenset, so that's not canonicalization in my view. Neither is type checking since a failed type check just leads to an error anyway. I also talked about this [in my earlier comment](https://github.com/dask/dask/issues/7933#issuecomment-895552688). 

It does sound though like two equivalent expressions would hash the same, so at least for that hashing would be usable for equality. 

> Edit: actually, I think we need to regenerate the token for cases like column projection, since failure to do so would break our hashing guarantees, allow semantically different tasks to still have identical keys. This would be bad. So instead we may need a way to replace static keys in custom layers with new names? Or make df.name always a uuid, and insert an alias during expr -> graph lowering to connect the staticly provided uuid with the dynamically generated hash.

It sounds to me like you conflating two separate issues, the hash as an identifier and the hash as equality. To me this suggests that they should indeed be separated.

**Comment by jcrist** [comment]
I think we're talking past each other, partly because our tower of abstractions here is complicated and partly because I'm doing a bad job of explaining things. I think this would be easier to clarify when we meet later this week.

**Comment by asmeurer** [comment]
I agree, although it would be helpful if I had something concrete to work on in the interim, if you can suggest anything. For now, the only "work" I feel I can do on this is trying to become more familiar with the current Dask internals, and trying to think about how this should be implemented based on my (likely confused) understanding of things.

**Comment by jcrist** [comment]
That's fair. I'm actually free now for a bit if you want to meet to go through the current dask internals.

**Comment by gerrymanoim** [comment]
Thanks for highlighting this thread in the Dask Heartbeat post. I wanted to mention that we're working on a similar-ish (more high level?) effort in https://github.com/ibis-project/ibis/, which now has a (very alpha) dask backend. Following the developments here with interest.

**Comment by asmeurer** [comment]
@gerrymanoim any references you can provide for similar work in ibis would be helpful to see how you are tackling the same problems. My understanding was that ibis already has an expression system, so I'd be interested to hear what works and what doesn't work in the ibis design.

**Comment by saulshanabrook** [comment]
Hey all,

I am reading through this thread to try to get up to speed with the current status of this work.

I have also been exploring some different ways of handling expression graphs in Python (previously in [metadsl](https://github.com/metadsl/metadsl) and now in a different soon-to-be open-sourced library) and am curious about exploring what approaches might work for Dask.

I know that currently in Dask there is a concept of a High Level Graph. This thread seems to suggest adding an alternative or additional system.

So I am wondering if anyone would be interested in summarizing:

1. What are some concrete examples of optimizations Dask would like to support but cannot due to its existing optimization strategies? (A parquest column loading example has been mentioned above a few times. Is this the main driver?)
2. Are there examples of optimizations that the current High Level Graph implements that should also be possible if a new system was developed that replaced it? Is the way that they are currently implemented with the High Level Graph not ideal for some reason?

Personally, I find it much easier to iterate on a solution if the constraints are well defined in terms of how it should perform, since the space of implementing expression systems is so large.

It's also exciting to hear that Ibis is considering restructuring its expressions as well! I think it would unlock a lot of potential in the ecosystem if libraries like Ibis and Dask could start sharing some of this machinery!

**Comment by rjzamora** [comment]
Thanks for continuing this discussion @saulshanabrook - I apologize in advance for the following information dump that is unlikely to answer any of your actual questions (since my immediate thoughts focus more on the high-level requirements of Dask rather than the requirements of the specific expression system that we ultimately need)…

I think you are correct that a clear list of required features and optimizations would be useful.  The column-projection example has been used in several places, because it a relatively simple optimization to understand.  I haven’t spent much time thinking through the other collection APIs, but I suspect that the primary optimizations we need in Dask-Dataframe are:

1. **partition culling** - Available for both dictionaries and HLGs. Fast and easy with HLGs.
2. **task fusion** - Available for both dictionaries and HLGs. Slow for large dictionaries. Faster for HLGs, but support is currently limited to adjacent `Blockwise`-based HLG layers.
3. **column projection** - Only available in HLGs. Support currently limited to `getitem` operations directly following a `DataframeIOLayer`.
4. **predicate pushdown** - Not yet available in Dask. Basic support is certainly possible with HLGs, but the distinct “partition-planning” stage used in read_parquet/read_orc makes this more complex than column projection.

Furthermore, my personal understanding of the “high-level” requirements for the overall HLG and/or HLE system is:

- **We want to avoid materializing a full dictionary-based graph when creating or operating on a collection**
    - This is already available with the existing HLG system, and is what has been the **focus** of that design (in my opinion).
- **We want to generate an optimized graph at `compute`/`persist` time**
    - This is already available with the existing HLG system for optimizations (1)-(3), but the code is already a mess without general support for (3) or basic support for (4).

From my perspective, the High-Level Expression conversation was motivated by a consensus that the HLG system (alone) is turning out to be an unnecessarily messy and painful way to achieve the desired high-level goals.  I think that we all agree that it should be possible to get what we need out of HLGs, but experience has clearly illuminated some HLG pain points:

- **”Proper” serialization is complicated**
    - The need to avoid importing pandas or numpy on the scheduler has made HLG-Layer implementations surprisingly complex. We have also been challenged by the need to avoid unpicking data on the scheduler.
- **It is difficult to use Information within the graph to optimize the graph**
    - Since the `HighLevelGraph` API was effectively implemented as a “drop-in” alternative for python dictionaries, the HLG does not track basic DataFrame-collection information (schema, partition count, etc…), and is eagerly generated before `compute`/`persist` is called.  This means that Dask cannot generate an optimal HLG from scratch, but must use the limited information stored within the “un-optimized” graph to produce an “optimized” graph. Optimizations (1)-(3) currently do this by modifying the original HLG.

In my opinion, it is this second pain point which makes a high-level expression system attractive. Adding a high-level expression object to the Dask collection seems unlikely to help us with serialization.  However, such a system *should* make optimization much-less “hacky.”  This is because we could avoid eager construction of a graph, and instead use this expression object to produce an optimized HLG at `compute`/`persist` time (or when an operation without HLE support is used).

**Comment by saulshanabrook** [comment]
Thank you @rjzamora for the very in-depth reply! The high-level requirements you listed are getting at exactly my original question.  Before we can determine the requirements of the expression system I think it's helpful to outline, as you have started, what are its intended use cases and the shortcomings of current solutions to those.

I want to take some time to digest the specific examples and pain points you listed above.

I was thinking a possible next step could be to come up with complete examples for each and pointers to the existing code which implements the optimizations, so we are all on the same page on the current "hackiness" and its repercussions for limitations in optimizing concrete use cases.

**Comment by eriknw** [comment]
I've been looking at HLEs again recently, so let me share at a very high level how I'd like to tackle it in a way that helps us experiment and learn.  I'm going to be somewhat vague and hand-wavy, because there is so much good information and details above that don't need repeating, and because I don't have a detailed, low level plan that I want to advocate for as the way forward (I actually have several in mind, including adapting prior efforts).

In principle, I believe some form (or perhaps many forms) of HLEs is achievable.  But, I expect any approach to be experimental for a while, and we're likely to want to iterate.  As others have pointed out, some of the difficulties in implementing HLEs in Dask are that the problem, design, solution spaces are high-dimensional, and that Dask has a large, complicated API surface area to cover, which may require some invasive changes.  I looked into HLEs briefly a few years ago, and I think it is daunting to try to experiment with it directly in the Dask codebase (but this may just be me).

My first goal is to create a sort of playground that makes it easier to experiment.  I envision this to be new Dask collections that mimic Array, DataFrame, and Bag collections (Delayed and user-defined collections can follow, but I'm not thinking about them yet).  By adding a shim layer, we can easily capture the function arguments, and call the methods on the original Dask collections when necessary.  Hopefully this will give us much of what we need, but I'm certain there will be shortcomings.  So, I'm also making it easy to modify the code of the original function.  There may be a little magic *\*cough cough\**, but the goal is to be convenient, flexible, and maintainable-ish (i.e., it should complain loudly when code it changes is changed).  I plan to package this effort into a, uh, Python package.  Maybe multiple packages for different experiments.  This should make it easier for other people to try it out, and to experiment themselves.  If invasive changes become too extensive or if maintenance becomes too much a burden, I expect it would be straightforward to convert such a package to a Dask PR.  Adding a layer around Dask isn't a new idea--mrocklin above said "The efforts I've played with / proposed in the past have been for layers surrounding Dask. That adds more complexity though"--but I think the complexity is worth it short term, and such a layer _may_ be worth it long term.  It may also make it easier to adapt to use for other packages, not just Dask.

Using MatchPy seems like a great idea.  I think we should.  There are many things we could match against and behaviors we could expose to the user.

My near term goal is to get a playground package that performs HLE pattern matching and rewriting for some DaskSQL patterns, including more robust predicate pushdown.  This should be like bread and butter for MatchPy.  From there, I think there will be many directions we could go or new experiments to try.

As a former boss of mine used to say, "for developers, doing is learning." There is a lot for me to learn (which is to say I have whole bunch of questions and concerns), and I'll get there by doing (hopefully--or I may crash and burn!).  I'll provide an update once I have a working playground.  I would also like to share other experiments I'd like to try.  Again, my main goal here is to learn, and to make it easier for other people to experiment and learn as well.  There's a lot of interest--and potential--in HLEs, so I say let's push forward however we can.

If anybody has a violent reaction to the way I'm thinking here, please reply (non-violently ;) ).

**Comment by saulshanabrook** [comment]
Hey @eriknw,

This sounds like an exciting plan!

I wanted to mention that I started with a similar intent a few years ago to make a library that makes it possible to make shim layers like this, called [metadsl](https://metadsl.readthedocs.io/en/latest/index.html). It actually originally used MatchPy under the hood to do the replacement, but over time I built up an alternative matcher and UX and ended up replacing it entirely. The main improvement it provides, in terms of UX, over constructing MatchPy expression directly, is the ability to describe your nodes as Python function and classes and build them up by calling them. This also lets you use type annotations to describe the relationship of the node types of the args and return values.

I went looking through the history and found this earlier version which creates MatchPy opreations dynamically from Python functions: https://github.com/metadsl/metadsl/blob/bee9ddbb3b5d31c7a5f42a0c4ea42f19ad45168e/uarray/machinery.py#L154-L173

I had originally designed it for another use case, but I will have more bandwidth soon and would be excited to chat with you about what I have learned from there and to see if I could help in anyway, if you would be interested!

**Comment by jcrist** [comment]
Hi @eriknw, happy to have someone else thinking about this issue. I looked at this a couple months ago and came to the following high-level conclusions (apologies for the unorganized nature of this comment):

- Any change made will have to be done incrementally, which leads to organizational issues. Even if we know where we want to end up (we don't), we'll also have to figure out a set of incremental changes we can make where each step is valid and doesn't break the world. This is hard for something as low level as the collections api.
- There are a bunch of complications that make it easier to have the expression tree (or whatever we call it) as an attribute on the user facing class, rather than the top-level class itself be an expression. By this I mean, the user would still only see an e.g. `dask.array.Array`, but an attribute on that class would contain the expression nodes/layers/whatever describing the computation to get to that point.
- If we're storing an expression tree on the class itself, why not just make high-level graphs better? They're already an expression system (just one that's hard to work with), and group pieces of the shared expression state together. It's easier to work incrementally from what we have already than to start a new place where state will live.

As such, I think we should work on improving the existing system by:
- Cleaning up how high-level-graphs generate low level graphs. Right now the code path for this is convoluted and coupled to the serialization logic (to avoid unpickling on the scheduler). @rjzamora has a PR up right now attempting to clean up some of this, but it'll be an iterative effort to make this simpler.
- Expanding what's stored on a `Layer` in the graph. Right now we don't store enough metadata on each layer to perform high level operations, but we can change that! We might move `_meta` and `divisions` to live on the layer for example.
- Deprecating and eventually removing the `Mapping` interface for things in the `.dask` attribute. Originally `.dask` was a real dict, then it was a `SharedDict`, then it was a `HighLevelGraph`. We're sufficiently divorced from the actual data structure looking like a mapping, and keeping the mapping interface is sometimes confusing and sometimes complicated. We'll still want to support layers in the graph composed of low-level dask dict representations, but at this point I think the high-level object on `.dask` shouldn't need to look like a `Mapping` anymore.

> Using MatchPy seems like a great idea. I think we should. There are many things we could match against and behaviors we could expose to the user.

IMO, the optimizations we want to write currently are few in number and custom enough that writing them by hand doesn't seem that onerous. Our main issue is not how to express the optimizations, it's more how do we store the expression tree and metadata in a way that lets us rewrite things before computing them:
- Efficiently
- Without breaking low level layers in the graph that might depend on fixed key names in upstream & downstream layers
- Without accidentally duplicating computations (there's a bunch of nuance to this, see e.g. https://github.com/dask/dask/issues/7933#issuecomment-918424353).

If after we solve all these and we end up in a situation where handwriting the optimizations is hard, then I'd call that a success (and maybe a time to adopt such a system). But we're so far from that point right now that I don't think it's worth thinking about.

Anyway, if you ever want to hop on a call to talk further, please let me know. Due to how much of our internals we've exposed as a public interface here, changes here have a ton of nuance and may seem fine for one operation and become a footgun for another. This is a complicated problem.

**Comment by caspervdw** [comment]
Hi all, I have been reading this issue with great interest. Slice pushdown in dask would be a pivotal feature for our application.

I (with my colleagues from Nelen & Schuurmans) developed [dask-geomodeling](https://github.com/nens/dask-geomodeling) a few years ago to address this issue specifically for the pushdown of geospatial requests. Using that package we store expression trees in a database and from that create task graphs on-the-fly after a user requested a subset through a web application. After we developed this, other projects like [dask-geopandas](https://github.com/geopandas/dask-geopandas) came up and we realized that we are missing out on a lot of features because we took an approach that basically reimplements every single operation. I am currently looking for ways to make more close use of `dask` directly. One must-have for us is this expression system.

So I am really excited reading the ideas here and the plans in https://github.com/dask/dask/issues/9159.

It may be useful to share some experiences we had running (or own style of) high level optimization in production for a couple of years.

### (de)serialization of expressions

At `dask-geomodeling` we use the same data structure for serializing layers as for serializing the tasks. Instead of functions, you use classes (types). The result of evaluating a layer graph is a collection instance ("view" in terms of dask-geomodeling). See https://dask-geomodeling.readthedocs.io/en/latest/views.html#view-serialization for an example.

I am no expert of ideal datastructures for optimization approaches. But the advantage here is that you can re-use code for things like constructing and optimization. Materializing the collection is like "computing". I tried to quickly create a mock implementation for this approach, but in the end I circled back and realized that `dask.delayed` does exactly this job:

```
import dask
import dask.dataframe as dd

ReadParquet = dask.delayed(dd.read_parquet)
src = ReadParquet(path)

def collection_optimize(dsk, keys):
    """ Your high-level optimizations go here """
    return dsk


with dask.config.set(delayed_optimize=collection_optimize):
    collection = src.compute()

collection.compute()  # computes the result
``` 

In my example above you can directly start solving the optimization problem (or use existing `RewriteRule` classes).
The "delayed" collection could be top-level or carried along as an attribute the the collections.

While I realize that this would be something "next to" HLG instead of an extension of HLGs (as described in #9159), I do think this approach is interesting for its simplicity. There are no new concepts introduced. Also I feel that HLGs are violating the "Single Responsibilty Principle" too much (but this is just a hunch that needs exploration).

There are cons here: one being that the `Delayed()` has to clue of the parameters and return type of the wrapped function. So there is no way for parameter parsing/validation or listing methods. You could do `src.non_existing_method(foo="bar")`. Also, there is no way for accessing metadata (like `columns`) for optimization rules.

I'd like to explore this option, but I'd like to know first what is your position on this approach. Maybe such an approach was already discarded, in that case sorry for the noise.

**Comment by rjzamora** [comment]
Thanks for sharing @caspervdw !

If I undertstand correctly, you are pointing out that much of the logic in `Delayed` can be reused to design a simple expression system for dask collections. This is a cool idea!  It does seem worth exploring - I'll try to think more about the possible opportunities and blockers.
